### PR TITLE
Add service level deployment errors when failing k8s api requests

### PR DESCRIPTION
**Pull request by kevo1ution**
## Changes proposed in this PR

- Provides a deployment error if there is an k8s api error
- Propagates message provided in the k8s api error response if it exists
- Example response to kitchen.localtest.me when K8s api request fails with message "unauthorized"
Screenshot from 2021-03-26 10-02-17

## Why are we making these changes?

- Users are provided info on why their service is failing

### CODE REVIEW

File: waiter/src/waiter/scheduler/kubernetes.clj [hunk 1]
```diff
@@ -429,8 +439,17 @@
...
-  [{:keys [watch-state] :as scheduler}]
-  (-> watch-state deref :service-id->service vals))
+  [{:keys [service-id->deployment-error-cache watch-state]}]
+  (let [service-id->service (-> watch-state deref :service-id->service)
+        service-id->deployment-error (cu/cache->map service-id->deployment-error-cache)
+        service-ids (set (concat (keys service-id->deployment-error) (keys service-id->service)))]
+    (map
+      (fn [service-id]
+        (let [deployment-error (get-in service-id->deployment-error [service-id :data])
+              service (get service-id->service service-id (create-empty-service service-id))]
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler/kubernetes.clj:449]:
```suggestion
              service (or (get service-id->service service-id) (create-empty-service service-id))]
```
Using `or` will lazily evaluate the second argument.

File: waiter/src/waiter/scheduler/kubernetes.clj [hunk 2]
```diff
@@ -608,11 +628,20 @@
...
-                                   :body (utils/clj->json rs-spec)
-                                   :request-method :post)
+        response-json
+        (ss/try+
+          (api-request request-url scheduler :body (utils/clj->json rs-spec) :request-method :post)
+          (catch Object response
+            (let [k8s-message (get-in response [:body :message] "Unknown reason - check logs")
+                  deployment-error (str "Failing K8s Request: " k8s-message)]
+              (log/info "creating deployment error for service" {:service-id service-id
+                                                                 :deployment-error deployment-error})
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler/kubernetes.clj:638]: Nitpick: lexicographic order

File: waiter/src/waiter/scheduler/kubernetes.clj [hunk 3]
```diff
@@ -608,11 +628,20 @@
...
+              (cu/cache-evict service-id->deployment-error-cache service-id)
+              (cu/cache-get-or-load service-id->deployment-error-cache service-id (constantly deployment-error))
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler/kubernetes.clj:640]: We should introduce a new `cache-utils` method to set the cache value.

File: waiter/src/waiter/service.clj [hunk 4]
```diff
@@ -262,7 +262,7 @@
                 instance)
               (if (and instance (not= instance :no-matching-instance-found))
                 ; instance is a deployment error if it (1) does not have an :id tag, (2) is not nil, and (3) does not equal :no-matching-instance-found
-                (ex-info (str "Deployment error: " (utils/message instance)) {:service-id service-id :status http-503-service-unavailable})
+                (ex-info (str "Deployment error: " (utils/message instance :default instance)) {:service-id service-id :status http-503-service-unavailable})
```
Comment (shamsimam) [inline waiter/src/waiter/service.clj:265]: Why is this necessary?

Comment (kevo1ution) [reply waiter/src/waiter/service.clj:265]: The instance can either be an instance or a deployment error message. (utils/message) function deserializes the instance into a proper deployment error message, but in this situation instance can be a k8s api error message. So if instance message is not found in deployment messages map, then we default to instance message itself.

Comment (shamsimam) [reply waiter/src/waiter/service.clj:265]: I would use this then:
```suggestion
                (ex-info (str "Deployment error: " (or (utils/message instance) instance)) {:service-id service-id :status http-503-service-unavailable})
```

File: waiter/src/waiter/scheduler/kubernetes.clj [hunk 5]
```diff
@@ -69,6 +70,17 @@
...
+(defn default-k8s-message-transform
+  "Takes k8s api error response and creates a user friendly deployment error message"
+  [k8s-error-response]
+  (get-in k8s-error-response [:body :message] "Unknown reason - check logs"))
+
+(defn create-service-deployment-error
+  "Transforms the deployment error message to be more user friendly"
+  [k8s-error-response response->deployment-error-msg-fn]
+  {:service-deployment-error-msg (response->deployment-error-msg-fn k8s-error-response)
+   :service-deployment-error-details {:k8s-response-body (get-in k8s-error-response [:body])}})
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler/kubernetes.clj:82]: nitpick: lexicographic order

File: waiter/src/waiter/util/cache_utils.clj [hunk 6]
```diff
@@ -55,6 +55,11 @@
       (.get key (fn cache-loader [] {:data (get-fn)}))
       :data))
 
+(defn cache-set
```
Comment (shamsimam) [inline waiter/src/waiter/util/cache_utils.clj:58]: Nitpick: `cache-put!`

Comment (shamsimam) [reply waiter/src/waiter/util/cache_utils.clj:58]: Please add unit tests for this.

**Review by shamsimam — DISMISSED** [review]

File: waiter/src/waiter/service.clj [hunk 7]
```diff
@@ -262,7 +262,16 @@
                 instance)
               (if (and instance (not= instance :no-matching-instance-found))
                 ; instance is a deployment error if it (1) does not have an :id tag, (2) is not nil, and (3) does not equal :no-matching-instance-found
-                (ex-info (str "Deployment error: " (utils/message instance)) {:service-id service-id :status http-503-service-unavailable})
+                (let [{:keys [service-deployment-error-msg service-deployment-error-details]} instance
+                      {:keys [error-message error-map]}
+                      (cond->
+                        {:error-message (utils/message instance)
```
Comment (shamsimam) [inline waiter/src/waiter/service.clj:268]: Nitpick: lexicographic order

**Review by shamsimam — APPROVED** [review]

File: waiter/src/waiter/scheduler/kubernetes.clj [hunk 8]
```diff
@@ -1395,6 +1435,8 @@
                       (utils/assoc-if-absent :client-name "waiter-k8s")
                       (utils/assoc-if-absent :user-agent "waiter-k8s")
                       hu/http-client-factory)
+        service-id->deployment-error-cache (cu/cache-factory {:threshold 5000
+                                                              :ttl (-> 5 t/minutes t/in-millis)})
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler/kubernetes.clj:1439]: The cache should be configurable

**Review by shamsimam — APPROVED** [review]
